state/options/tx_options.go

```diff
@@ -0,0 +1,201 @@
+package options
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math"
+	"strings"
+
+	"github.com/cosmos/cosmos-sdk/crypto/keyring"
+	sdktypes "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/celestiaorg/celestia-app/pkg/user"
+	apptypes "github.com/celestiaorg/celestia-app/x/blob/types"
+)
+
+const (
+	// gasMultiplier is used to increase gas limit in case if tx has additional options.
+	gasMultiplier = 1.1
+
+	//	Since 0 is a valid fee input for the Tx, the default value is -1.
+	defaultFeeAmount = -1
+)
+
+var (
+	errNoGasProvided     = errors.New("gas limit was not set")
+	errNoAddressProvided = errors.New("address was not set")
+)
+
+// TxOptions specifies additional options that will be applied to the Tx.
+type TxOptions struct {
+	// fee is private since it has to be set through `SetFeeAmount`
+	fee      *Fee
+	GasLimit uint64
+
+	// Specifies the key from the keystore associated with an account that
+	// will be used to sign transactions.
+	// NOTE: This `Account` must be available in the `Keystore`.
+	Account string
+	// Specifies the account that will pay for the transaction.
+	// Input format Bech32.
+	Granter string
+}
+
+func DefaultTxOptions() *TxOptions {
+	return &TxOptions{
+		fee: DefaultFee(),
+	}
+}
+
+type jsonTxOptions struct {
+	Fee      *Fee   `json:"fee,omitempty"`
+	GasLimit uint64 `json:"gasLimit,omitempty"`
+	Account  string `json:"account,omitempty"`
+	Granter  string `json:"granter,omitempty"`
+}
+
+func (options *TxOptions) MarshalJSON() ([]byte, error) {
+	jsonOpts := &jsonTxOptions{
+		Fee:      options.fee,
+		GasLimit: options.GasLimit,
+		Account:  options.Account,
+		Granter:  options.Granter,
+	}
+	return json.Marshal(jsonOpts)
+}
+
+func (options *TxOptions) UnmarshalJSON(data []byte) error {
+	var jsonOpts jsonTxOptions
+	err := json.Unmarshal(data, &jsonOpts)
+	if err != nil {
+		return fmt.Errorf("unmarshalling TxOptions:%w", err)
+	}
+
+	options.fee = jsonOpts.Fee
+	options.GasLimit = jsonOpts.GasLimit
+	options.Account = jsonOpts.Account
+	options.Granter = jsonOpts.Granter
+	return nil
+}
+
+// SetFeeAmount sets fee for the transaction.
+func (options *TxOptions) SetFeeAmount(amount int64) {
+	if amount >= 0 {
+		options.fee.Amount = amount
+		options.fee.isSet = true
+	}
+}
```

Comment:
This method is for go users right? Because if I'm submitting a JSON RPC request, then the `Fee` field is public and I would fill it out myself.

Reply:
Yes. It will be handled in `UnmarshalJSON`.